### ISSUE TITLE
Support artifact servers that return "unauthorised" instead of "not found"

### ISSUE THREAD

**Issue by FrancescoCapari** [gradle/gradle#3335]
### Expected Behavior
Gradle should resolve dependencies from custom repositories without errors.

### Current Behavior
Gradle fails to resolve dependencies because of breaking changes introduced in 4.3 that you can not control if you are using JitPack.io.

### Context
I'm sure that this issue gets closed without a resolution (as the other two issues reported recently did), but I thought I could spare some time for those users having the same issue as I do - so you don't have to bother about upgrading to Gradle 4.3 and then going back to 4.2.1 that works correctly.

If you are using JitPack and / or have multiple repositories, **do not upgrade to Gradle 4.3** otherwise you would break your builds. When you have certain dependency in your own repository and Gradle 4.3 tries to find it in JitPack, your build fails. You can not do anything about it and re-ordering the repositories in your build script (as suggested by developers in another thread) is **not** an acceptable solution.

### Steps to Reproduce (for bugs)
There are tens of dependencies that can trigger this. Just add random library that does not exists in JitPack, e.g. https://jitpack.io/com/github/random/Random/1.0/Random-1.0.pom

### Your Environment
You will get "Received status code 401 from server: Unauthorized" from JitPack and build fails even thought you have this library in your own repository.

**TL;DR [WARNING] do not upgrade to Gradle 4.3 if you are using multiple repositories or JitPack.io, otherwise you would not be able to build your projects.**

**Comment by oehme** [comment]
This is a problem in jitpack. It should return `404`, not `401` when something isn't found.

Why is reordering not acceptable?

**Comment by FrancescoCapari** [comment]
1) Because the order should not matter and it can be part of library we're not authors of.
2) Does not make any difference, the 401 error is thrown no matter the order of the repositories.

**Comment by FrancescoCapari** [comment]
@oehme How exactly is throwing an exception when *any* repository returns 4xx client error (but 404) for *any* dependency hiding problems in a build configuration?

**Comment by mark-vieira** [comment]
> This is a problem in jitpack. It should return 404, not 401 when something isn't found.

I'm not sure I agree. In fact, this behavior with GitHub is actually problematic in some cases and requires the use of preemptive auth. I wouldn't say returning a 404 for what is effectively a permissions issue is idomatic API design, albeit common.

> I think it's reasonable that Gradle fails on authorisation errors. Otherwise we would be hiding actual problems in your build configuration.

So this comes down to this change: https://docs.gradle.org/current/release-notes.html#avoid-checking-other-repositories-when-dependency-resolution-in-one-repository-fails

I think your logic is reasonable, the whole purpose behind this change was that failure to access repositories could effectively not surface due to ordering, etc, causing different dependencies to be used in different build invocations. However, this is in itself potentially a use-case, relying on repository ACLs to determine which dependencies for a build to use. It's a bit goofy but not entirely unreasonable. I don't see how ordering would affect this, you'd have to programmatically "omit" the repositories you don't have access to.

I'm slightly leaning towards us relaxing this restriction to exclude client errors as @FrancescoCapari has suggested. This is a pretty disruptive change for folks relying on this behavior. I certainly agree that server errors (5XX) or other IO exceptions (timeout, etc) should trigger the build to fail but perhaps a good middle ground is proceeding on client errors. In almost every case, folks that rely completely on auth for their repositories will still encounter a failed build as the required dependencies simply won't be found.

**Comment by danlange** [comment]
Another scenario:
We have artifactory configured to have separate snapshot and release repositories.  If you request a snapshot from a release repository, it will return a 409 conflict response, to let you know you are looking in the wrong repository.
In 4.2.1, this was not a problem, but in 4.3, gradle sees the 409 response and throws an exception.
I agree with @mark-vieira that all 4xx responses should be treated the same way as 404s.

**Comment by FrancescoCapari** [comment]
@mark-vieira @danlange Thanks for your contribution. Agreed - all HTTP client errors (4xx) should not fail a dependency lookout.

Since dropping Gradle would be quite complicated for us and since we're afraid that this breaking change (quite literally) won't get fixed, we currently at least wrote a simple proxy server that simply searches for dependencies and returns 401 for all client errors - this is the only way we could think of to *make Gradle great again* and usable for us. Quite ridiculous solution, but at least it works.

**Comment by oehme** [comment]
I'm not sure ignoring *all* 4xx codes is the way to go (e.g. think of "Unsupported Method") , but I get the point.

On the technical level we can treat 401 and 409 the same as 404 (i.e. remember that the module was missing and continue to the next repo), though we should probably present a different error to the user if the module couldn't be found at all. I.e. instead of "not found" we should spell out "not authenticated" or "not applicable".

**Comment by mark-vieira** [comment]
I think handling individual client error codes differently is fraught with danger. In practice, these error codes are implemented to mean very different things depending on the server. We can't reliable take any individual error code to mean anything other than "the server didn't like the question we asked". I don't see much benefit in cherry picking some codes to bail on and others to proceed. Like I've said before, in most of these scenarios the build will fail anyhow.

**Comment by oehme** [comment]
Let's back the change out for 4.3.1 and add it back in with a more lenient behavior in 4.4.

**Comment by bmuschko** [comment]
@bigdaz It think this issue has been fixed by your PRs https://github.com/gradle/gradle/pull/3381 and https://github.com/gradle/gradle/pull/3382. Can we close this?

**Comment by oehme** [comment]
Yes, though we should open a follow-up to handle 5xx better.

**Comment by bmuschko** [comment]
@oehme [Created issue](https://github.com/gradle/gradle/issues/3398) and closing this one.

**Comment by bmuschko** [comment]
@FrancescoCapari This has been fixed. Could you try out your project with a snapshot of the 4.3.1 release? https://downloads.gradle.org/distributions-snapshots/gradle-4.3.1-20171107130851+0000-bin.zip

**Comment by FrancescoCapari** [comment]
@bmuschko **BUILD SUCCESSFUL** (I tried your snapshot and I can verify that this issue is no longer reproducible with 4.3.1 snapshot). Thank you guys.

**Comment by bmuschko** [comment]
@FrancescoCapari With Gradle 4.4 we will treat dependency resolution handling for HTTP 5xx error codes the same way as we do for connection and socket timeouts (do not ask subsequent repositories). we'd love for you to try it out and provide us with feedback to avoid any potential issues in the final release. You can try out a snapshot distribution here: https://services.gradle.org/distributions-snapshots/gradle-4.4-20171112235945+0000-bin.zip. The actual change can be found here if you are interested: https://github.com/gradle/gradle/pull/3412. thanks for your support! let us know if you have any questions.
